assembly/src/parsers/ast/serialization.rs

```diff
@@ -0,0 +1,1165 @@
+                target.write_felt(*v);
+            }
+            Self::MemStoreW => target.write_opcode(OpCode::MemStoreW),
+            Self::MemStoreWImm(v) => {
+                target.write_opcode(OpCode::MemStoreWImm);
+                target.write_felt(*v);
+            }
+            Self::LocStoreW(v) => {
+                target.write_opcode(OpCode::LocStoreW);
+                target.write_felt(*v);
+            }
+            Self::LoadWAdv => target.write_opcode(OpCode::LoadWAdv),
+
+            Self::PushConstants(constants) => {
+                target.write_opcode(OpCode::PushConstants);
+                let length = constants.len();
+                target.write_u8(length as u8);
+                for v in constants {
+                    target.write_felt(*v);
+                }
+            }
+            Self::AdvU64Div => target.write_opcode(OpCode::AdvU64Div),
+            Self::AdvPush(v) => {
+                target.write_opcode(OpCode::AdvPush);
+                target.write_felt(*v);
+            }
+            Self::AdvLoadW(v) => {
+                target.write_opcode(OpCode::AdvLoadW);
+                target.write_felt(*v);
+            }
+
+            // ----- cryptographic operations ---------------------------------------------------------
+            Self::RPHash => target.write_opcode(OpCode::RPHash),
+            Self::RPPerm => target.write_opcode(OpCode::RPPerm),
+            Self::MTreeGet => target.write_opcode(OpCode::MTreeGet),
+            Self::MTreeSet => target.write_opcode(OpCode::MTreeSet),
+            Self::MTreeCWM => target.write_opcode(OpCode::MTreeCWM),
+
+            // ----- exec / call ----------------------------------------------------------------------
+            Self::ExecLocal(v) => {
+                target.write_opcode(OpCode::ExecLocal);
+                target.write_u32(*v);
+            }
+            Self::ExecImported(imported) => {
+                target.write_opcode(OpCode::ExecImported);
+                target.write_string(imported);
+            }
+            Self::CallLocal(v) => {
+                target.write_opcode(OpCode::CallLocal);
+                target.write_u32(*v);
+            }
+            Self::CallImported(imported) => {
+                target.write_opcode(OpCode::CallImported);
+                target.write_string(imported);
+            }
+        }
+    }
+}
+
+pub struct ByteReader;
+
+impl ByteReader {
+    pub fn get_bool(bytes: &mut &[u8]) -> Result<bool, SerializationError> {
+        let (result, remaining_bytes) = bytes.split_at(1);
+        *bytes = remaining_bytes;
+        let result = u8::from_le_bytes(result.try_into().unwrap());
+        u8_to_bool(result)
+    }
+
+    pub fn get_u8(bytes: &mut &[u8]) -> Result<u8, SerializationError> {
+        let (result, remaining_bytes) = bytes.split_at(1);
+        *bytes = remaining_bytes;
+        let result = u8::from_le_bytes(result.try_into().unwrap());
+
+        Ok(result)
+    }
```

Comment:
I would not have 2 different variants of function for `get` vs. `read`. I think we should be able to just have `read` which returns a value of a specific type and also advances the position pointer.